mono/sgen/sgen-marksweep.c

```diff
@@ -1818,6 +1818,9 @@ major_start_major_collection (void)
 			sweep_block (block);
 		SGEN_ASSERT (0, block->state == BLOCK_STATE_SWEPT, "All blocks must be swept when we're pinning.");
 		set_block_state (block, BLOCK_STATE_MARKING, BLOCK_STATE_SWEPT);
+		/* Blocks that have a null free_list are full. Evacuating them is pointless. */
```

Comment:
Of course there will be an impact, but that is not because evacuation was more effective before, but because we were behaving as a copying collector for these blocks. Following this logic, we could just do evacuation on every block on all collections, and we would have a copying collector and it would lead to better compaction. I'm also curios to see some more numbers.